File: src/cmd/info.zig
const std = @import("std");
const tool_mod = @import("../tool.zig");
const state_mod = @import("../state.zig");
const output = @import("../ui/output.zig");

const help =
    \\Usage: dot info <tool>
    \\
    \\Show detailed information about a tool.
    \\
    \\Options:
    \\  --help, -h    Show this help
    \\
    \\Examples:
    \\  dot info kubectl
    \\  dot info helm
    \\
;

pub fn run(
    allocator: std.mem.Allocator,
    args: []const []const u8,
    state: *state_mod.State,
    tools: []const tool_mod.Tool,
) !void {
    for (args) |a| {
        if (std.mem.eql(u8, a, "--help") or std.mem.eql(u8, a, "-h")) {
            output.printRaw(help);
            return;
        }
    }

    if (args.len == 0) {
        output.printRaw(help);
        return;
    }

    const query = args[0];

    // Find tool by id or alias
    var found: ?*const tool_mod.Tool = null;
    for (tools) |*t| {
        if (std.mem.eql(u8, t.id, query)) { found = t; break; }
        for (t.aliases) |a| {
            if (std.mem.eql(u8, a, query)) { found = t; break; }
        }
        if (found != null) break;
    }

    if (found == null) {
        output.printUnknownTool(query);
        return;
    }

    const t = found.?;
    const entry = state.tools.get(t.id);

    // Header: "==> name — description"
    var hdr_buf: [256]u8 = undefined;
    const hdr = std.fmt.bufPrint(&hdr_buf, "{s} — {s}", .{ t.name, t.description }) catch t.name;
    output.printSectionHeader(hdr);

    // Details block
    output.printFmt("\n", .{});
    output.printFmt("  {s}Homepage:{s}     {s}\n", .{ output.bold, output.reset, t.homepage });

    // Groups
    var grp_buf: [64]u8 = undefined;
    var gw: std.Io.Writer = .fixed(&grp_buf);
    for (t.groups, 0..) |g, i| {
        if (i > 0) gw.writeByte(',') catch {};
        gw.writeAll(@tagName(g)) catch {};
    }
    const grp = gw.buffered();
    if (grp.len > 0) {
        output.printFmt("  {s}Groups:{s}       {s}\n", .{ output.bold, output.reset, grp });
    }

    // Aliases
    if (t.aliases.len > 0) {
        var alias_buf: [64]u8 = undefined;
        var aw: std.Io.Writer = .fixed(&alias_buf);
        for (t.aliases, 0..) |a, i| {
            if (i > 0) aw.writeAll(", ") catch {};
            aw.writeAll(a) catch {};
        }
        output.printFmt("  {s}Aliases:{s}      {s}\n", .{ output.bold, output.reset, aw.buffered() });
    }

    // Status section
    output.printSectionHeader("Status");
    output.printFmt("\n", .{});

    if (entry) |e| {
        const home = @import("../env.zig").getenv("HOME") orelse "";
        var bin_buf: [512]u8 = undefined;
        const bin_path = std.fmt.bufPrint(&bin_buf, "{s}/.local/bin/{s}", .{ home, t.id }) catch "";
        const pin_str: []const u8 = if (e.pinned) "yes" else "no";
        output.printFmt("  {s}Installed:{s}    {s}{s}{s}\n", .{ output.bold, output.reset, output.green, e.version, output.reset });
        output.printFmt("  {s}Binary:{s}       {s}\n", .{ output.bold, output.reset, bin_path });
        if (e.installed_at.len > 0) {
            var date_buf: [24]u8 = undefined;
            const date = output.fmtTimestamp(e.installed_at, &date_buf);
            output.printFmt("  {s}Installed at:{s} {s}\n", .{ output.bold, output.reset, date });
        }
        output.printFmt("  {s}Method:{s}       {s}\n", .{ output.bold, output.reset, e.method });
        output.printFmt("  {s}Pinned:{s}       {s}\n", .{ output.bold, output.reset, pin_str });

        // Resolve latest version (non-fatal)
        const latest = t.version_source.resolve(allocator) catch null;
        defer if (latest) |l| allocator.free(l);
        if (latest) |l| {
            if (std.mem.eql(u8, l, e.version)) {
                output.printFmt("  {s}Latest:{s}       {s}up to date{s}\n", .{ output.bold, output.reset, output.green, output.reset });
            } else {
                output.printFmt("  {s}Latest:{s}       {s} {s}(update available){s}\n", .{ output.bold, output.reset, l, output.yellow, output.reset });
            }
        }
    } else {
        output.printFmt("  {s}not installed{s}\n", .{ output.dim, output.reset });
        output.printFmt("\n  Run '{s}dot install {s}{s}' to install.\n", .{ output.bold, t.id, output.reset });
    }

    // Quick start
    if (t.quick_start.len > 0) {
        output.printSectionHeader("Quick Start");
        output.printFmt("\n", .{});
        for (t.quick_start) |line| {
            output.printFmt("  {s}\n", .{line});
        }
    }

    // Resources
    if (t.resources.len > 0) {
        output.printSectionHeader("Resources");
        output.printFmt("\n", .{});
        for (t.resources) |r| {
            output.printFmt("  {s}{s}{s}: {s}\n", .{ output.bold, r.label, output.reset, r.url });
        }
    }

    output.printFmt("\n", .{});
}

File: src/cmd/info.rs
use crate::state::State;
use crate::tool::Tool;
use crate::ui::output;
use clap::Args;
use console::style;

#[derive(Debug, Args)]
pub struct InfoArgs {
    /// Tool ID to show info for
    pub tool: String,
}

pub fn run(args: &InfoArgs, state: &State, tools: &[Tool]) -> anyhow::Result<()> {
    let query = &args.tool;

    // Find by id or alias
    let found = tools
        .iter()
        .find(|t| t.id == query.as_str() || t.aliases.iter().any(|a| a == query));

    let t = match found {
        Some(t) => t,
        None => {
            output::print_unknown_tool(query);
            // Suggest closest
            if let Some(suggestion) = closest_tool(query, tools) {
                eprintln!("Did you mean '{suggestion}'?");
            }
            return Ok(());
        }
    };

    let colored = output::get_render_mode() == output::RenderMode::Rich;
    let bold_label = |s: &str| -> String {
        if colored {
            style(s).bold().to_string()
        } else {
            s.to_string()
        }
    };

    // Header: "Name - Description" (or just "Name" if no description)
    if t.description.is_empty() {
        output::print_section_header(&t.name);
    } else {
        output::print_section_header(&format!("{} - {}", t.name, t.description));
    }
    eprintln!();
    if !t.homepage.is_empty() {
        eprintln!("  {}     {}", bold_label("Homepage:"), t.homepage);
    }

    let groups_str = t
        .groups
        .iter()
        .map(|g| g.name())
        .collect::<Vec<_>>()
        .join(", ");
    if !groups_str.is_empty() {
        eprintln!("  {}       {groups_str}", bold_label("Groups:"));
    }

    if !t.aliases.is_empty() {
        eprintln!("  {}      {}", bold_label("Aliases:"), t.aliases.join(", "));
    }

    // Status section
    output::print_section_header("Status");
    eprintln!();

    let entry = state.get_entry(&t.id);
    if let Some(e) = entry {
        let home = dirs::home_dir().unwrap_or_default();
        let bin_path = home.join(".local/bin").join(&t.id);

        let version_str = if colored {
            style(&e.version).green().to_string()
        } else {
            e.version.clone()
        };
        eprintln!("  {}    {version_str}", bold_label("Installed:"));
        eprintln!("  {}       {}", bold_label("Binary:"), bin_path.display());

        if !e.installed_at.is_empty() {
            let date = output::fmt_timestamp(&e.installed_at);
            eprintln!("  {} {date}", bold_label("Installed at:"));
        }
        eprintln!("  {}       {}", bold_label("Method:"), e.method);
        eprintln!(
            "  {}       {}",
            bold_label("Pinned:"),
            if e.pinned { "yes" } else { "no" }
        );

        // Resolve latest version (non-fatal)
        if let Ok(latest) = t.version_source.resolve() {
            if latest == e.version {
                let up_str = if colored {
                    style("up to date").green().to_string()
                } else {
                    "up to date".to_string()
                };
                eprintln!("  {}       {up_str}", bold_label("Latest:"));
            } else {
                let avail_str = if colored {
                    format!("{latest} {}", style("(update available)").yellow())
                } else {
                    format!("{latest} (update available)")
                };
                eprintln!("  {}       {avail_str}", bold_label("Latest:"));
            }
        }
    } else {
        eprintln!("  not installed");
        eprintln!("\n  Run 'dot install {}'  to install.", t.id);
    }

    // Quick start
    if !t.quick_start.is_empty() {
        output::print_section_header("Quick Start");
        eprintln!();
        for line in &t.quick_start {
            eprintln!("  {line}");
        }
    }

    // Resources
    if !t.resources.is_empty() {
        output::print_section_header("Resources");
        eprintln!();
        for r in &t.resources {
            let label_str = if colored {
                style(&r.label).bold().to_string()
            } else {
                r.label.clone()
            };
            eprintln!("  {label_str}: {}", r.url);
        }
    }

    eprintln!();
    Ok(())
}

fn closest_tool<'a>(query: &str, tools: &'a [Tool]) -> Option<&'a str> {
    use crate::util::edit_distance;
    const THRESHOLD: usize = 3;
    let mut best_dist = usize::MAX;
    let mut best_id = None;
    for t in tools {
        let d = edit_distance(&t.id, query);
        if d < best_dist {
            best_dist = d;
            best_id = Some(t.id.as_str());
        }
    }
    if best_dist <= THRESHOLD {
        best_id
    } else {
        None
    }
}
